crates/net/discv5/src/lib.rs

```diff
@@ -0,0 +1,789 @@
+    }
+
+    /// Applies filtering rules on an ENR. Returns [`Ok`](FilterOutcome::Ok) if peer should be
+    /// passed up to app, and [`Ignore`](FilterOutcome::Ignore) if peer should instead be dropped.
+    fn filter_discovered_peer(&self, enr: &discv5::Enr) -> FilterOutcome {
+        self.discovered_peer_filter.filter(enr)
+    }
+
+    /// Returns the [`ForkId`] of the given [`Enr`](discv5::Enr), if field is set.
+    fn get_fork_id<K: discv5::enr::EnrKey>(
+        &self,
+        enr: &discv5::enr::Enr<K>,
+    ) -> Result<ForkId, Error> {
+        let mut fork_id_bytes = enr.get_raw_rlp(self.fork_id_key()).ok_or(Error::ForkMissing)?;
+
+        Ok(ForkId::decode(&mut fork_id_bytes)?)
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////
+    // Interface with sigp/discv5
+    ////////////////////////////////////////////////////////////////////////////////////////////////
+
+    /// Exposes API of [`discv5::Discv5`].
+    pub fn with_discv5<F, R>(&self, f: F) -> R
+    where
+        F: FnOnce(&Self) -> R,
+    {
+        f(self)
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////
+    // Complementary
+    ////////////////////////////////////////////////////////////////////////////////////////////////
+
+    /// Returns the [`IpMode`] of the local node.
+    pub fn ip_mode(&self) -> IpMode {
+        self.ip_mode
+    }
+
+    /// Returns the key to use to identify the [`ForkId`] kv-pair on the [`Enr`](discv5::Enr).
+    pub fn fork_id_key(&self) -> &[u8] {
+        self.fork_id_key
+    }
+}
+
+impl fmt::Debug for Discv5 {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        "{ .. }".fmt(f)
+    }
+}
+
+/// Result of successfully processing a peer discovered by [`discv5::Discv5`].
+#[derive(Debug)]
+pub struct DiscoveredPeer {
+    /// A discovery v4 backwards compatible ENR.
+    pub node_record: NodeRecord,
+    /// [`ForkId`] extracted from ENR w.r.t. configured
+    pub fork_id: Option<ForkId>,
+}
+
+/// Gets the next lookup target, based on which distance is currently being targeted.
+pub fn get_lookup_target(
+    log2_distance: usize,
+    local_node_id: discv5::enr::NodeId,
+) -> discv5::enr::NodeId {
+    let mut target = local_node_id.raw();
+    //make sure target has a 'distance'-long suffix that differs from local node id
+    if log2_distance != 0 {
+        let suffix_bit_offset = MAX_LOG2_DISTANCE.saturating_sub(log2_distance);
+        let suffix_byte_offset = suffix_bit_offset / 8;
+        // todo: flip the precise bit
+        // let rel_suffix_bit_offset = suffix_bit_offset % 8;
+        target[suffix_byte_offset] = !target[suffix_byte_offset];
+
+        if suffix_byte_offset != 31 {
+            for b in target.iter_mut().take(31).skip(suffix_byte_offset + 1) {
+                *b = rand::random::<u8>();
```

Comment:
can you explain why this is needed? is this specced somewhere?

Reply:
not specced, it's an addition based on experience. it's better than looking up a random node id in queries that have aim to populate kbuckets. for a random node id, there is 50% chance the node id falls in the bucket furthest away. as requested distances is capped at 3 distances, the likelihood that any peers discovered as part of a query, will fall in the top 3 kbuckets is 50%. that means all peers discovered in all queries, are competing over the same 16 * 3 places in the kbuckets, where 16 is the max nodes per kbuckets. for health of the network we want to be connected to every subtree in the network.
<img width="484" alt="Screenshot 2024-04-01 at 22 04 56" src="https://github.com/paradigmxyz/reth/assets/58548332/75a815fe-0ee6-4110-bab7-ce35810144d5">

Reply:
note: if all apps using discv5 did this, DAS would be a piece of cake, which is about finding specific node ids, because nodes that are online for a longer time, would have nicely populated kbuckets (not only the 3 buckets furthest away). I think we want to keep this bigger perspective for what's beneficial for whole ethereum now when we enter discv5 universe :)